keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,1096 @@
+
+#### Using label selectors
+
+We consider using label selectors by the Job controller to identify the subset
+of jobs it should watch. However, due to backwards-compatibility there are two
+cases:
+1. old jobs without the label
+2. new jobs with the label equal to `job-controller`
+
+Using label selectors might be relevant in case of [Story 1](#story-1), as
+MultiKueue can have many jobs in the management cluster. Keeping the jobs in
+the Job controller's informer, and starting the `syncJob` function for them is
+wasteful.
+
+However, the label selector semantics does not support OR-ing these two cases.
+For this reason, we need to ensure all jobs have the label set. One mechanism
+to consider is to use the `defaults.go` file to set the label. However, this
+might have a drawback, that it will require many writes to etcd after a cluster
+upgrade. This concern can be addressed, to some extent, by incubating the Beta
+version which sets the default for new jobs, for a release or more, before
+using `defaults.go`. Then, it seems reasonable to assume that only a small
+portion of jobs would require an actual etcd write after cluster upgrade.
+
+We add evaluation of the strategy to [GA requirements](#ga).
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+#### Ecosystem fragmentation due to forks
+
+The mechanism to disable the main Job controller opens the door for users to
+substitute it with a fork. This may create more fragmentation in the community
+as users may prefer to use their forked controllers rather than contribute
+upstream.
+
+First, this risk, to some extent, exists even today as admins with access to the
+control plane can disable by passing `controllers=-job,*` in the manifest for of
+`kube-controller-manager`.
+
+Second, we believe that users who had the need to fork the Job controller
+already introduced dedicated Job CRDs for their needs.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+#### API
+
+```golang
+const (
+  ...
+	// LabelManagedBy is used to indicate the controller or entity that manages
+	// an Job.
+	LabelManagedBy = "batch.kubernetes.io/managed-by"
```

Comment:
I like the simplicity of the surface area here. Some benefits of the label approach over a field:
* it generalizes to custom resources as well
* it could consistently live in the same place in multiple objects (all objects have metadata.labels)

The main downside I see is that there's no signal in openapi whether the cluster is new enough to be aware of the controller subdivision. That's largely a problem time would cure.

I wonder if it would make sense to teach all the core workload controllers a generic version of this (`kubernetes.io/managed-by`) on their parent objects. That would enable use of multiple controllers on the same type, with the default (no managed-by label) meaning "use the default controller for this type".

Reply:
Thank you!

> The main downside I see is that there's no signal in openapi whether the cluster is new enough to be aware of the controller subdivision. That's largely a problem time would cure.

Indeed, I added a note on the related risk in the  [Two controllers running at the same time on old version](https://github.com/kubernetes/enhancements/blob/e2b8c22bd2912d61b4f99b4ae5aa94f2b044ada7/keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md#two-controllers-running-at-the-same-time-on-old-version) section

Reply:
> I wonder if it would make sense to teach all the core workload controllers a generic version of this (`kubernetes.io/managed-by`) on their parent objects. That would enable use of multiple controllers on the same type, with the default (no managed-by label) meaning "use the default controller for this type".

@atiratree this ^ would probably be a good goal for sig-apps long term to address.

Reply:
It is an interesting idea. I suppose we could consider it, if there were clear benefits and a clearly defined process and contract for the ownership.

The downsides mentioned in this KEP are still quite significant to support this in all controllers. I think CRDs are still preferable for extending workloads like Deployments, StatefulSets, etc.

I know we have conformance tests, but that doesn't stop people from forking the controllers and implementing incompatible behavior. And API validation has its limits. This would then be problematic for the consumers of these workloads who would either be unprepared and unable to react in a safe and bug-free manner. And if they could, they would need extra logic and switch the behavior according to the current controller. They would also run into a similar problem if they found a controller they do not support.

As we see for example with StatefulSets and their need for migration. I expect the same need to arise for these controllers. E.g. I want to migrate from a controller A to a controller B for a new shiny functionality. So sooner or later we might make the managed label/field mutable. This would probably require some kind of locking, so that two controller would not operate on the same object. Even in mid change. And to consider rogue controllers.

The benefit for Jobs is that it is used as a building block in many projects, and we do not expect people to fork the controller. Not sure if this applies elsewhere.

@mimowo can we add the following to the GA graduation criteria?

Asses the fragmentation of the ecosystem. Look for other implementations of a job controller and asses their conformance with k8s.

Reply:
Makes sense, added the point.

Reply:
> Regarding the generic `kubernetes.io/managed-by` label. I would like to defer it for this KEP.

Generic is definitely outside the scope of this KEP.